### ISSUE TITLE
Error Parsing IP Adress when trying to stream

### ISSUE THREAD

**Issue by CC84** [MPExtended/MPExtended#394]
When trying to stream liveTV to my PC, nothing happens. So i saved the m3u file and tried to open it in VLC, getting the error: 
Ihre Eingabe konnte nicht geöffnet werden:
VLC kann die MRL 'http://cc-htpc/Stream/TV?item=905&transcoder=Direct&continuationId=playlist-42958' nicht öffnen. Sehen Sie für Details im Fehlerprotokoll nach.

I tried to open this URL in Firefox and get the attached error.

MP is a SingleSeat Installation running under Win8.
WebMediaPortal is running in IIS.

Client is Windows 7.

Streaming to my Win8 Netbook and also on the server itself works fine. So maybe it's an Problem between Win8 und Win7, don't know. Unfortunatly i uninstalled VisualStudio some days ago, maybe you are able to find the error anyway, otherways contact me and i'll debug in it and check to local vars.

```
WebMediaPortal version 0.5.3 (commit df8b4b0).
Request: http://cc-htpc/Stream/TV?item=905&transcoder=Direct&continuationId=playlist-42958
```

Der Index war außerhalb des Arraybereichs.

   bei MPExtended.Libraries.Service.Network.IPAddressExtensions.GetNetworkAddress(IPAddress address, IPAddress subnetMask) in C:\Users\Oxan\Documents\Projects\MPExtended\Libraries\MPExtended.Libraries.Service\Network\IPAddressExtensions.cs:Zeile 30.
   bei MPExtended.Libraries.Service.Network.IPAddressExtensions.IsInSameSubnet(IPAddress address, IPAddress check, IPAddress subnetMask) in C:\Users\Oxan\Documents\Projects\MPExtended\Libraries\MPExtended.Libraries.Service\Network\IPAddressExtensions.cs:Zeile 42.
   bei MPExtended.Libraries.Service.Network.NetworkInformation.IsOnLAN(IPAddress address) in C:\Users\Oxan\Documents\Projects\MPExtended\Libraries\MPExtended.Libraries.Service\Network\NetworkInformation.cs:Zeile 103.
   bei MPExtended.Applications.WebMediaPortal.Controllers.StreamController.GetStreamMode() in C:\Users\Oxan\Documents\Projects\MPExtended\Applications\MPExtended.Applications.WebMediaPortal\Controllers\StreamController.cs:Zeile 122.
   bei MPExtended.Applications.WebMediaPortal.Controllers.StreamController.GenerateStream(WebMediaType type, String itemId, String transcoder, Int32 starttime, String continuationId) in C:\Users\Oxan\Documents\Projects\MPExtended\Applications\MPExtended.Applications.WebMediaPortal\Controllers\StreamController.cs:Zeile 185.
   bei lambda_method(Closure , ControllerBase , Object[] )
   bei System.Web.Mvc.ActionMethodDispatcher.Execute(ControllerBase controller, Object[] parameters)
   bei System.Web.Mvc.ReflectedActionDescriptor.Execute(ControllerContext controllerContext, IDictionary`2 parameters)
   bei System.Web.Mvc.ControllerActionInvoker.InvokeActionMethod(ControllerContext controllerContext, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
   bei System.Web.Mvc.ControllerActionInvoker.<>c__DisplayClass15.<InvokeActionMethodWithFilters>b__12()
   bei System.Web.Mvc.ControllerActionInvoker.InvokeActionMethodFilter(IActionFilter filter, ActionExecutingContext preContext, Func`1 continuation)
   bei System.Web.Mvc.ControllerActionInvoker.<>c__DisplayClass15.<>c__DisplayClass17.<InvokeActionMethodWithFilters>b__14()
   bei System.Web.Mvc.ControllerActionInvoker.InvokeActionMethodWithFilters(ControllerContext controllerContext, IList`1 filters, ActionDescriptor actionDescriptor, IDictionary`2 parameters)
   bei System.Web.Mvc.ControllerActionInvoker.InvokeAction(ControllerContext controllerContext, String actionName)
   bei System.Web.Mvc.Controller.ExecuteCore()
   bei System.Web.Mvc.ControllerBase.Execute(RequestContext requestContext)
   bei System.Web.Mvc.ControllerBase.System.Web.Mvc.IController.Execute(RequestContext requestContext)
   bei System.Web.Mvc.MvcHandler.<>c__DisplayClass6.<>c__DisplayClassb.<BeginProcessRequest>b__5()
   bei System.Web.Mvc.Async.AsyncResultWrapper.<>c__DisplayClass1.<MakeVoidDelegate>b__0()
   bei System.Web.Mvc.Async.AsyncResultWrapper.<>c__DisplayClass8`1.<BeginSynchronous>b__7(IAsyncResult _)
   bei System.Web.Mvc.Async.AsyncResultWrapper.WrappedAsyncResult`1.End()
   bei System.Web.Mvc.MvcHandler.<>c__DisplayClasse.<EndProcessRequest>b__d()
   bei System.Web.Mvc.SecurityUtil.<GetCallInAppTrustThunk>b__0(Action f)
   bei System.Web.Mvc.SecurityUtil.ProcessInApplicationTrust(Action action)
   bei System.Web.Mvc.MvcHandler.EndProcessRequest(IAsyncResult asyncResult)
   bei System.Web.Mvc.MvcHandler.System.Web.IHttpAsyncHandler.EndProcessRequest(IAsyncResult result)
   bei System.Web.HttpApplication.CallHandlerExecutionStep.System.Web.HttpApplication.IExecutionStep.Execute()
   bei System.Web.HttpApplication.ExecuteStep(IExecutionStep step, Boolean& completedSynchronously)

**Comment by CC84** [comment]
It took me some time, to debug this issue, but finally i could fix it.
I was not able to remote Debug the w3wp process with VS 2010 (there seems to be some problem with the VS 2010 remote Debugger under Win8), so i updated the solution to VS2012, deployed it and used the VS2012 Remote Debugger.

The problem i found is an IPv6/IPv4 problem. (You already added an Todo to this :))

When i try to stream to my Win7 PC, the Method IsOnLan is called with an IPv6 address. So you call IsInSameSubnet on an IPv6 Address, this one is 16 Byte long, but the IPv4Mask is only 4 Byte long, so in GetNetworkAddress the broadcastAddress is 16 byte long, but the subnetMaskBytes[] is only 4 bytes long and and the IndexOutOfBoundExpception is thrown.

I simply fixed this by adding this line to IsInSameSubnet:
if(address.IsIPv6LinkLocal) return false;

**Comment by oxan** [comment]
Something that throws me off: your error message says it tries to access an invalid array index in IPAddressExtensions.cs on line 30, but if I look at [that file in v0.5.3](/MPExtended/MPExtended/blob/v0.5.3/Libraries/MPExtended.Libraries.Service/Network/IPAddressExtensions.cs#L30), there's no array access there at all.

Anyway, your explanation makes sense. I try my best to support IPv6, but it's poorly-tested at best... I'll fix it for 0.5.4.

**Comment by CC84** [comment]
Am 22.06.2013 01:01, schrieb Oxan van Leeuwen:

> Something that throws me off: your error message says it tries to 
> access an invalid array index in IPAddressExtensions.cs on line 30, 
> but if I look at that file in v0.5.3 
> </MPExtended/MPExtended/blob/v0.5.3/Libraries/MPExtended.Libraries.Service/Network/IPAddressExtensions.cs#L30>, 
> there's no array access there at all.
> 
> Anyway, your explanation makes sense. I try my best to support IPv6, 
> but it's poorly-tested at best... I'll fix it for 0.5.4.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/MPExtended/MPExtended/issues/394#issuecomment-19845337.
> 
> Forget about the linenumber. It's wrong, don't know why. I even wasn't 
> able to debug the dlls and pdbs from your package. I had to build them 
> on my own.  The correct line was 35, like described below, i got it to 
> work with the described fix, but only for a short time. No again i went 
> into problems, without any error. In the logs there are the following lines:

2013-06-20 14:01:57.63726 [(15)] DEBUG: Starting stream type=TV; 
itemId=905; transcoder=Direct; starttime=0; continuationId=playlist-16007
2013-06-20 14:01:57.63726 [(15)] DEBUG: Stream is for user  from host 
fe80::bc14:47a2:4f76:182f%12, has identifier webmediaportal-64851 and is 
using mode Proxied with timeout 300s
2013-06-20 14:01:58.74487 [(15)] DEBUG: Stream started successfully and 
is at 
http://192.168.34.23:4322/MPExtended/StreamingService/stream/RetrieveStream?identifier=webmediaportal-64851
2013-06-20 14:01:58.74487 [(15)] DEBUG: Proxying stream from 
http://192.168.34.23:4322/MPExtended/StreamingService/stream/RetrieveStream?identifier=webmediaportal-64851 
with buffer size 65536
2013-06-20 14:01:58.93207 [(15)] WARN: Exception while proxying stream 
http://192.168.34.23:4322/MPExtended/StreamingService/stream/RetrieveStream?identifier=webmediaportal-64851
System.Web.HttpException (0x800704CD): Der Remotehost hat die Verbindung 
beendet. Fehlercode: 0x800704CD.
    bei 
System.Web.Hosting.IIS7WorkerRequest.RaiseCommunicationError(Int32 
result, Boolean throwOnDisconnect)
    bei System.Web.Hosting.IIS7WorkerRequest.ExplicitFlush()
    bei System.Web.HttpResponse.Flush(Boolean finalFlush, Boolean async)
    bei System.Web.HttpResponse.Flush()
    bei System.Web.HttpWriter.WriteFromStream(Byte[] data, Int32 offset, 
Int32 size)
    bei System.Web.HttpResponseStream.Write(Byte[] buffer, Int32 offset, 
Int32 count)
    bei 
MPExtended.Applications.WebMediaPortal.Controllers.StreamController.ProxyStream(String 
sourceUrl)
2013-06-20 14:01:58.93207 [(15)] DEBUG: Finished stream webmediaportal-64851

Seems to be again an IPv6 Problem. I solved this by disabling the IPv6 
Protocol on my HTPC, no further problems :)

I suspect, you have to do a lot of rework to support IPv6.

Anyway your coding style is realy good, i didn't had much problems to 
understand your code and your projects are well structured.

Thanks for this great project. If you need further support for testing, 
just ask.

Greetings

**Comment by oxan** [comment]
Interesting, I've seen that error a lot more, but never in relation to IPv6. The error simply indicates that the client has disconnected from the server, and isn't something we can fix in MPExtended. At best we can fix the reason why the client disconnects, but I really don't have an idea why it happens. 

> Anyway your coding style is realy good, i didn't had much problems to understand your code and your projects are well structured.

Thanks, that's always great to hear.

**Comment by CC84** [comment]
Am 22.06.2013 17:38, schrieb Oxan van Leeuwen:

> Interesting, I've seen that error a lot more, but never in relation to 
> IPv6. The error simply indicates that the client has disconnected from 
> the server, and isn't something we can fix in MPExtended. At best we 
> can fix the reason why the client disconnects, but I really don't have 
> an idea why it happens.
> 
> ```
> Anyway your coding style is realy good, i didn't had much problems
> to understand your code and your projects are well structured.
> ```
> 
> Thanks, that's always great to hear.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/MPExtended/MPExtended/issues/394#issuecomment-19858688.
> 
> Even in my company many workers are not able to setup a solution which 
> builds right after getting it from sourcecontrol. :)

But back to the problem: Maybe it was just a coincidence, cos it worked 
also with IPv6 for a short time. I'll try to reproduce this and watch 
about the communication on the line (wireshark), maybe this will give us 
some more information.

**Comment by oxan** [comment]
> Even in my company many workers are not able to setup a solution which builds right after getting it from sourcecontrol. :)
> Ah yes, that's one of my pet peeves. 

In any case, if you can reproduce it and watch the communication, that'd be very much appreciated.

**Comment by CC84** [comment]
> In any case, if you can reproduce it and watch the communication, 
> that'd be very much appreciated.
> 
> It's some time ago since i've used wirekshark. So i had to read into TCP 
> Protocol to understand the captured results, hope i'm right with my 
> interpretation.

First of all: It's reproducable on IPv6. Currently i assume that it 
never worked on IPv6. Maybe it's a kind of randomness if my local 
hostname is resolved to its IPv4 or IPv6 adress and the times it worked 
the IPv4 address was used.

When using IPv4 everything is fine, when using IPv6 the connections is 
canceled by the client after the first ~65000Bytes.

It's not easy to intpret the send data. The data should be the 
compressed video, which is not human readable, but i think i can 
identify single packets. Everytime a tcp Packet has not it's maximum 
size of 1514Bytes a packet should end. In IPv4 and IPv6 the first packet 
is about 65000Bytes long. Further packets (Only in IPv4) are much smaller.
Eventually the first IPv6 packet is not valid, which causes VLC to drop 
this stream, but as long as i'm not able to validate the transmitted 
data, it'll be hard to evaluate this theory.

Are you firm in using Wireshark? I can send you some captured logs if 
you tell me your mailadress.

Tomorrow i'll try to write a simple program, which saves the stream to 
disk. I've already tried to save the IPv6 stream by firefox, but in this 
case the stream stops after 124Bytes, while the IPv4 Stream runs endless 
as expected.

**Comment by CC84** [comment]
Am 23.06.2013 19:54, schrieb Oxan van Leeuwen:

> In any case, if you can reproduce it and watch the communication, 
> that'd be very much appreciated.
> 
> —
> Reply to this email directly or view it on GitHub 
> https://github.com/MPExtended/MPExtended/issues/394#issuecomment-19878042.
> 
> Maybe i've found the mistake. At least currently it's running.

Further i was able to reproduce this bug also with IPv4. The problem 
occurs with using your proxiing system. I forced my IPv4 connection to 
use this proxy as well, by modifing the IsInSameSubnet method to run 
false in all cases. Once my connections was forced to use the proxy it 
stopped working immediatly.

The next try was to connect to the stream by a simple .net tool with a 
WebRequest.

In wireshark i was able to see, the server was sending lots of data, but 
the response was always 124Bytes long.  The same 124Bytes Firefox saves 
to disk when saving the stream result to disk.

So i took a look to the HTTP Headers:

without Proxy:

HTTP/1.1 200 OK
Transfer-Encoding: chunked
Content-Type: video/x-unknown
Server: Microsoft-HTTPAPI/2.0
X-Content-Length: 124
Date: Sun, 23 Jun 2013 20:11:47 GMT
Connection: close

Proxy:
HTTP/1.1 200 OK
Cache-Control: private
Content-Length: 124
Content-Type: video/x-unknown
Server: Microsoft-IIS/8.0
X-AspNetMvc-Version: 3.0
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
Date: Sun, 23 Jun 2013 19:53:18 GMT
Connection: close

The Content-Length forces the client to take only the first 124 Bytes, 
the rest is dropped and the connection is dropped by the client.

The reason why i'm not sure, if my fix is correct. Your Proxymethod 
changes explicitly the X-Content-Length from the underlying stream to 
Content-Length, so i assume there is a good reason you have added this 
behavior. I removed the .Substring(2) call and now i can stream using 
the proxy method.

I'm not firm enough about http, but logically it should be wrong to add 
a Content-Length to a stream whose length is not known.

If you have further questions, just ask.

Greetings

**Comment by oxan** [comment]
Thanks, that is a very interesting and useful analysis. It's awesome to have bug reports who actually know what they're talking and can provide a detailed analysis of the problem.

The problem isn't related to WebMediaPortal at all, but WebMP is the only client that triggers it. As you discovered, we set the Content-Length header in the proxy system. This is done because the proxy system is also used for downloading files and sending playlists, both cases where it's useful to have a Content-Length header. The reason we copy it from the X-Content-Length header in the original response is that, due to (I think) a bug in WCF, we can't set the Content-Length header in that response. So the wrong value comes from the original response from the service.

Now, there are only two cases where the service sets the X-Content-Length header: when downloading a file (not relevant here), and when streaming a file using the Direct profile (as I expected we would always know the length of the stream in that case, since no transcoding is done). If you feel like it, you can verify it by reverting your fix and trying one of the Flash profiles, that should work (don't feel obligated, I've already verified it myself). However, as you said, the length of the stream isn't known in one case: when we're streaming live TV (you could try to stream a movie or recording, that should work). Due to live TV streams internally being represented by the path to their timeshifting index file (.tsbuffer.ts file in MediaPortal's timeshifting directory), the length of the timeshifting index file was erroneously set in the X-Content-Length header instead of omitting the header.

This is probably the same bug as #317, for which I had no idea why it was happening because it seemed to happen quite randomly. This bug indeed happens only when you (A) use the proxying system, (B) are streaming live TV and (C) are using the Direct profile. The combinations of these factors being a bit rare and the totally non-descriptive error message probably made the bug appear random to me.

Again, thanks very much for your analysis. It's very much appreciated. I'll push a fix for the root cause in a few hours.

**Comment by CC84** [comment]
I was me pleasure.

I know how annoying it is to get bugreports whose give you no hint how 
to reproduce them. By the way, even on my system it's still kind of 
randomnes. Don't know why, but my system still switches between using 
the IPv4 and the IPv6 adress randomly.

For me it was very satisfying to be able to localize this bug in foreign 
code. Even if you have a good codestyle, it's always a challenge, to be 
able to get a foreign solution to work (again a commendation for your 
work) and further to be able to find the right position. Even though i 
finally did not find the correct mistake, i was close to und additionaly 
was aware about the fact, the criticized code had to have some useful 
function.

I have to use direct streaming, cause my CPU (AMD E-350) is not able to 
reencode FullHD fast enough :)

I have not enough time to work seriously at such a project, but from 
time to time, i like the challenge.

If you have further "misterious" bugs, just tell me, i'll keep an eye on 
them.

**Comment by Ollienator** [comment]
Hi Oxan,

I think there still is an issue in the current implementation of LoadIPAddress() in NetworkInformation.cs.  Actually, there are two:
1. The following line is localization dependant:
`string[] preferedInterfaces = new string[] { "Local Area Connection" };`
On my system, the "Local Area Connection" interface name actually returns "LAN-Verbindung", so the preferred interface never gets picked and it falls through the next part of the code
2. This is more important: The first interface IP Address returned by 
            `if (addresses.Any()) return addresses.First();`
on my system is an IPv6 address.  So BaseAddresses.GetForService() tries to create an URI for that IPv6 address.  The problem is that the correct format for an IPv6 URI is "http://[address]:port/path" (note the square brackets) while in the current implementation the square brackets are missing and the new URI() call throws an exception.
I fixed it on my system by changing 
 `.Select(x => x.ToString())`
to
 `.Select(x => x.AddressFamily == AddressFamily.InterNetworkV6 ? "[" + x.ToString() + "]" : x.ToString())`
in LoadIPAddress() which is ugly, but it works.  I'm sure there is is a better way though,

**Comment by oxan** [comment]
Thanks; should be solved in respectively ef76305 and 06483d2.
